File: day4/python/day4_part1.py
with open("../input.txt") as f:
    grid = [line.rstrip('\n') for line in f if line.strip()]

rows = len(grid)
cols = len(grid[0]) if rows > 0 else 0

# 8 directions: all adjacent cells
directions = [(-1, -1), (-1, 0), (-1, 1),
              (0, -1),          (0, 1),
              (1, -1),  (1, 0), (1, 1)]

accessible = 0

for r in range(rows):
    for c in range(cols):
        if grid[r][c] != '@':
            continue

        # Count adjacent rolls
        neighbor_rolls = 0
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '@':
                neighbor_rolls += 1

        # Accessible if fewer than 4 adjacent rolls
        if neighbor_rolls < 4:
            accessible += 1

print(accessible)

File: day4/c/day4_part1.c
#include <stdio.h>
#include <string.h>

#define MAX_ROWS 150
#define MAX_COLS 150

static char grid[MAX_ROWS][MAX_COLS];
static int rows, cols;

/* 8 directions */
static int dr[] = {-1, -1, -1,  0, 0,  1, 1, 1};
static int dc[] = {-1,  0,  1, -1, 1, -1, 0, 1};

static int
count_neighbors(r, c)
int r, c;
{
	int count, i, nr, nc;

	count = 0;
	for (i = 0; i < 8; i++) {
		nr = r + dr[i];
		nc = c + dc[i];
		if (nr >= 0 && nr < rows && nc >= 0 && nc < cols)
			if (grid[nr][nc] == '@')
				count++;
	}
	return count;
}

static int
is_accessible(r, c)
int r, c;
{
	if (grid[r][c] != '@')
		return 0;
	return count_neighbors(r, c) < 4;
}

int
main()
{
	FILE *f;
	char line[256];
	int len, r, c, count;

	f = fopen("../input.txt", "r");
	if (f == NULL) {
		perror("../input.txt");
		return 1;
	}

	rows = 0;
	while (fgets(line, sizeof(line), f) != NULL) {
		len = strlen(line);
		if (len > 0 && line[len - 1] == '\n')
			len--;
		if (len == 0)
			continue;
		memcpy(grid[rows], line, len);
		cols = len;
		rows++;
	}
	fclose(f);

	count = 0;
	for (r = 0; r < rows; r++)
		for (c = 0; c < cols; c++)
			if (is_accessible(r, c))
				count++;

	printf("Part 1: %d\n", count);
	return 0;
}
